Platform/OpenNtfsDxe/Index.c

```diff
@@ -0,0 +1,938 @@
+        }
+      } else {
+        GoUpALevel (Context);
+      }
+
+      Status = FindFile (Symlink, Context);
+      FreePool (Symlink);
+      if (EFI_ERROR (Status)) {
+        return Status;
+      }
+    }
+  }
+
+  DEBUG ((DEBUG_INFO, "NTFS: File `%s' not found\n", IterCtx.Name));
+  return EFI_NOT_FOUND;
+}
+
+EFI_STATUS
+FsHelpFindFile (
+  IN  CONST CHAR16     *Path,
+  IN  NTFS_FILE        *RootNode,
+  OUT NTFS_FILE        **FoundNode,
+  IN  FSHELP_FILETYPE  Type
+  )
+{
+  EFI_STATUS      Status;
+  FSHELP_CTX      Context;
+  FSHELP_FILETYPE FoundType;
+
+  ASSERT (Path != NULL);
+  ASSERT (RootNode != NULL);
+  ASSERT (FoundNode != NULL);
+
+  Context.Path         = Path;
+  Context.RootNode     = RootNode;
+  Context.SymlinkDepth = 0;
+  Context.CurrentNode  = NULL;
+
+  if (Path[0] != L'/') {
+    DEBUG ((DEBUG_INFO, "NTFS: Invalid file name `%s'\n", Path));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = GoToRoot (&Context);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = FindFile ((CHAR16 *)Path, &Context);
+  if (EFI_ERROR (Status)) {
+    FreeStack (&Context);
+    return Status;
+  }
+
+  *FoundNode = Context.CurrentNode->Node;
+  FoundType = Context.CurrentNode->Type;
+  //
+  // Do not free the node
+  //
+  Context.CurrentNode->Node = NULL;
+  FreeStack (&Context);
+
+  if (FoundType != Type) {
+    if (*FoundNode != RootNode) {
+      FreeFile (*FoundNode);
+      FreePool (*FoundNode);
+    }
+
+    if (Type == FSHELP_REG) {
+      DEBUG ((DEBUG_INFO, "NTFS: Not a regular file\n"));
+    } else if (Type == FSHELP_DIR) {
+      DEBUG ((DEBUG_INFO, "NTFS: Not a directory\n"));
+    }
+
+    Status = EFI_VOLUME_CORRUPTED;
+  }
+
+  return Status;
+}
+
+EFI_STATUS
+IterateDir (
+  IN NTFS_FILE     *Dir,
+  IN VOID          *FileOrCtx,
+  IN FUNCTION_TYPE FunctionType
+  )
+{
+  EFI_STATUS          Status;
+  NTFS_ATTR           Attr;
+  ATTR_HEADER_RES     *Res;
+  ATTR_HEADER_NONRES  *Non;
+  ATTR_INDEX_ROOT     *Index;
+  INDEX_RECORD_HEADER *IndexRecord;
+  UINT8               *BitIndex;
+  UINT8               *BitMap;
+  UINTN               BitMapLen;
+  UINT8               Bit;
+  UINTN               Number;
+
+  ASSERT (Dir != NULL);
+  ASSERT (FileOrCtx != NULL);
+
+  if (!Dir->InodeRead) {
+    Status = InitFile (Dir, Dir->Inode);
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+  }
+
+  IndexRecord = NULL;
+  BitMap = NULL;
+
+  Status = InitAttr (&Attr, Dir);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+  //
+  // Search in $INDEX_ROOT
+  //
+  while (TRUE) {
+    Res = (ATTR_HEADER_RES *) FindAttr (&Attr, AT_INDEX_ROOT);
+    if (Res == NULL) {
+      DEBUG ((DEBUG_INFO, "NTFS: no $INDEX_ROOT\n"));
+      FreeAttr (&Attr);
+      return EFI_VOLUME_CORRUPTED;
+    }
+
+    mBufferSize = mFileRecordSize - (Attr.Current - Attr.BaseMftRecord->FileRecord);
+
+    if ((mBufferSize < sizeof (*Res))
+      || (mBufferSize < (Res->NameOffset + 8))
```

Comment:
Better to assign all these numbers more meaningful names?

Reply:
Do you think that `StrSize (L"$I30") - sizeof (CHAR16)` or `StrLen (L"$I30") * sizeof (CHAR16)` will look better?
Is meaning of 8 not obvious in `CompareMem ((UINT8 *) Res + Res->NameOffset, L"$I30", 8)`?